engine/access/rest/websockets/controller.go

```diff
@@ -36,66 +38,77 @@ func NewWebSocketController(
 func (c *Controller) HandleConnection(ctx context.Context) {
 	//TODO: configure the connection with ping-pong and deadlines
 	//TODO: spin up a response limit tracker routine
-	go c.readMessagesFromClient(ctx)
-	c.writeMessagesToClient(ctx)
+	defer c.shutdownConnection()
+	go c.readMessages(ctx)
+	c.writeMessages(ctx)
 }
 
-// writeMessagesToClient reads a messages from communication channel and passes them on to a client WebSocket connection.
+// writeMessages reads a messages from communication channel and passes them on to a client WebSocket connection.
 // The communication channel is filled by data providers. Besides, the response limit tracker is involved in
 // write message regulation
-func (c *Controller) writeMessagesToClient(ctx context.Context) {
-	//TODO: can it run forever? maybe we should cancel the ctx in the reader routine
+func (c *Controller) writeMessages(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
-		case msg := <-c.communicationChannel:
+		case msg, ok := <-c.communicationChannel:
+			if !ok {
+				return
+			}
+			c.logger.Debug().Msgf("read message from communication channel: %s", msg)
+
 			// TODO: handle 'response per second' limits
 
 			err := c.conn.WriteJSON(msg)
 			if err != nil {
+				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) ||
+					websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+					return
+				}
+
 				c.logger.Error().Err(err).Msg("error writing to connection")
+				return
 			}
+
+			c.logger.Debug().Msg("written message to client")
 		}
 	}
 }
 
-// readMessagesFromClient continuously reads messages from a client WebSocket connection,
+// readMessages continuously reads messages from a client WebSocket connection,
 // processes each message, and handles actions based on the message type.
-func (c *Controller) readMessagesFromClient(ctx context.Context) {
-	defer c.shutdownConnection()
-
+func (c *Controller) readMessages(ctx context.Context) {
 	for {
-		select {
-		case <-ctx.Done():
-			c.logger.Info().Msg("context canceled, stopping read message loop")
-			return
-		default:
-			msg, err := c.readMessage()
-			if err != nil {
-				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseAbnormalClosure) {
-					return
-				}
-				c.logger.Warn().Err(err).Msg("error reading message from client")
+		msg, err := c.readMessage()
+		if err != nil {
+			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseAbnormalClosure) ||
+				websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
 				return
 			}
 
-			baseMsg, validatedMsg, err := c.parseAndValidateMessage(msg)
-			if err != nil {
-				c.logger.Debug().Err(err).Msg("error parsing and validating client message")
-				return
-			}
+			c.logger.Debug().Err(err).Msg("error reading message from client")
+			continue
```

Comment:
what type of errors are we allowing here? does it make sense to just close the connection?